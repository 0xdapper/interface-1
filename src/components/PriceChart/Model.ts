import { Token } from '@uniswap/sdk-core'
import { scaleLinear } from 'd3-scale'
import { curveBasis, line } from 'd3-shape'
import { useMemo } from 'react'
import { parse } from 'react-native-redash'
import { GraphData, GraphMetadatas, PriceList } from 'src/components/PriceChart/types'
import { useDailyTokenPrices, useHourlyTokenPrices } from 'src/features/historicalChainData/hooks'
import { dimensions } from 'src/styles/sizing'
import { logger } from 'src/utils/logger'

export const HEIGHT = 180
export const WIDTH = dimensions.fullWidth

export const NUM_GRAPHS = 5

const HOURS_IN_DAY = 24
const HOURS_IN_WEEK = 24 * 7
const HOURS_IN_MONTH = HOURS_IN_WEEK * 4
const DAYS_IN_YEAR = 365

// TODO(#89): use date manipulation util
const d = new Date()
const ONE_MONTH_AGO = d.setMonth(d.getMonth() - 1)

export function useGraphs(token: Token): GraphMetadatas | null {
  const hourlyTokenData = useHourlyTokenPrices({ token, timestamp: ONE_MONTH_AGO })
  const dailyTokenData = useDailyTokenPrices({ token })

  const isError = dailyTokenData.isError || hourlyTokenData.isError

  const hourlyTokenPrices = useMemo(
    () => ({
      // TODO(judo): interpolation requires same data length
      // TODO(#80): use block-level data and add 1h chart
      oneDay: hourlyTokenData.prices?.slice(0, HOURS_IN_DAY),
      oneWeek: hourlyTokenData.prices?.slice(0, HOURS_IN_WEEK),
      oneMonth: hourlyTokenData.prices?.slice(0, HOURS_IN_MONTH),
    }),
    [hourlyTokenData]
  )

  const dailyTokenPrices = useMemo(
    () => ({
      oneYear: dailyTokenData.prices?.slice(0, DAYS_IN_YEAR),
      all: dailyTokenData.prices,
    }),
    [dailyTokenData]
  )

  return useMemo(() => {
    if (isError) {
      logger.error(
        'PriceChart/Model',
        'useGraphs',
        'Historical prices error',
        dailyTokenData.error ?? hourlyTokenData.error
      )
      return null
    }

    const graphs = [
      {
        label: '1D',
        index: 0,
        data: buildGraph(hourlyTokenPrices.oneDay),
      },
      {
        label: '1W',
        index: 1,
        data: buildGraph(hourlyTokenPrices.oneWeek),
      },
      {
        label: '1M',
        index: 2,
        data: buildGraph(hourlyTokenPrices.oneMonth),
      },
      {
        label: '1Y',
        index: 3,
        data: buildGraph(dailyTokenPrices.oneYear),
      },
      {
        label: 'all',
        index: 4,
        data: buildGraph(dailyTokenPrices.all),
      },
    ] as const

    return graphs as GraphMetadatas
  }, [
    isError,
    hourlyTokenPrices.oneDay,
    hourlyTokenPrices.oneWeek,
    hourlyTokenPrices.oneMonth,
    dailyTokenPrices.oneYear,
    dailyTokenPrices.all,
    dailyTokenData.error,
    hourlyTokenData.error,
  ])
}

/**
 * Constructs a drawable Path from a PriceList
 */
export function buildGraph(datapoints: PriceList | undefined): GraphData | null {
  if (!datapoints || datapoints.length === 0) return null

  const priceList = datapoints.slice(0, 24).reverse()

  const formattedValues = priceList.map(
    (price) => [price.timestamp, parseFloat(price.close)] as [number, number]
  )

  const prices = formattedValues.map(([, price]) => price)
  const dates = formattedValues.map(([date]) => date)

  const minPrice = Math.min(...prices)
  const maxPrice = Math.max(...prices)
  const startingPrice = prices[0]

  const scaleX = scaleLinear()
    .domain([dates[0], dates[dates.length - 1]])
    .range([0, WIDTH])
  const scaleY = scaleLinear().domain([minPrice, maxPrice]).range([HEIGHT, 0])

  const path = parse(
    line()
      .x(([x]) => scaleX(x) as number)
      .y(([, y]) => scaleY(y) as number)
      .curve(curveBasis)(formattedValues) as string
  )

  return {
    minPrice,
    maxPrice,
    startingPrice,
    path,
  }
}
