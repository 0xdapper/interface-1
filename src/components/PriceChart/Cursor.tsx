import React from 'react'
import { StyleSheet } from 'react-native'
import { PanGestureHandler } from 'react-native-gesture-handler'
import {
  useAnimatedGestureHandler,
  useAnimatedStyle,
  useSharedValue,
  withSpring,
} from 'react-native-reanimated'
import { getYForX } from 'react-native-redash'
import { AnimatedBox, Box } from 'src/components/layout/Box'
import { AnimatedIndex, AnimatedTranslation, GraphMetadatas } from 'src/components/PriceChart/types'

const CURSOR_SIZE = 50

interface CursorProps {
  index: AnimatedIndex
  translation: AnimatedTranslation
  graphs: GraphMetadatas
}

export const Cursor = ({ index, translation, graphs }: CursorProps) => {
  const isActive = useSharedValue(false)
  const onGestureEvent = useAnimatedGestureHandler({
    onStart: () => {
      isActive.value = true
    },
    onActive: (event) => {
      translation.x.value = event.x
      translation.y.value = getYForX(graphs[index.value].data.path, translation.x.value) || 0
    },
    onEnd: () => {
      //TODO(judo): reset cursor
      isActive.value = false
    },
  })

  const style = useAnimatedStyle(() => {
    const translateX = translation.x.value - CURSOR_SIZE / 2
    const translateY = translation.y.value - CURSOR_SIZE / 2
    return {
      transform: [{ translateX }, { translateY }, { scale: withSpring(isActive.value ? 1 : 0) }],
    }
  })

  return (
    <Box flex={1} style={StyleSheet.absoluteFill}>
      <PanGestureHandler onGestureEvent={onGestureEvent}>
        <AnimatedBox style={StyleSheet.absoluteFill}>
          <AnimatedBox
            alignItems="center"
            borderRadius="full"
            height={CURSOR_SIZE}
            justifyContent="center"
            style={[styles.cursor, style]}
            width={CURSOR_SIZE}>
            <Box width={15} height={15} borderRadius="full" bg="primary1" />
          </AnimatedBox>
        </AnimatedBox>
      </PanGestureHandler>
    </Box>
  )
}

const styles = StyleSheet.create({
  cursor: {
    backgroundColor: 'rgba(243, 71, 191, 0.171)',
  },
})
