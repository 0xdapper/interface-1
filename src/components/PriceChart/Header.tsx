import { createRestyleComponent, createVariant, useTheme, VariantProps } from '@shopify/restyle'
import React from 'react'
import { interpolate, useAnimatedStyle, useDerivedValue } from 'react-native-reanimated'
import { ReText, round } from 'react-native-redash'
import { Box } from 'src/components/layout/Box'
import { HEIGHT } from 'src/components/PriceChart/Model'
import { AnimatedIndex, AnimatedTranslation, GraphMetadatas } from 'src/components/PriceChart/types'
import { Theme } from 'src/styles/theme'

interface HeaderProps {
  translation: AnimatedTranslation
  index: AnimatedIndex
  graphs: GraphMetadatas
}

// const StyledReText = createBox<Theme,  & typeof ReText.arguments>(ReText)

const StyledReText = createRestyleComponent<
  VariantProps<Theme, 'textVariants'> & React.ComponentProps<typeof ReText>,
  Theme
>([createVariant({ themeKey: 'textVariants' })], ReText)

export const Header = ({ translation, index, graphs }: HeaderProps) => {
  const theme = useTheme<Theme>()

  const data = useDerivedValue(() => graphs[index.value].data)

  const price = useDerivedValue(() =>
    interpolate(translation.y.value, [0, HEIGHT], [data.value.maxPrice, data.value.minPrice])
  )
  const percentChange = useDerivedValue(
    () => ((price.value - data.value.startingPrice) / data.value.startingPrice) * 100
  )

  const priceFormatted = useDerivedValue(() => {
    return `$${round(price.value, 2).toLocaleString('en-US', { currency: 'USD' })}`
  })
  const percentChangeFormatted = useDerivedValue(() => `${round(percentChange.value, 3)}%`)

  const style = useAnimatedStyle(() => ({
    color: percentChange.value > 0 ? theme.colors.green : theme.colors.red,
  }))

  return (
    <Box flex={1} padding="md" justifyContent="center">
      <Box flex={1} alignItems="center">
        <StyledReText variant="h1" text={priceFormatted} />
        <StyledReText variant="h3" style={style} text={percentChangeFormatted} />
      </Box>
    </Box>
  )
}
