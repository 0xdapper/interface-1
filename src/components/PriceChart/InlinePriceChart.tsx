import { Currency } from '@uniswap/sdk-core'
import dayjs from 'dayjs'
import React, { useMemo } from 'react'
import { serialize } from 'react-native-redash'
import Svg, { Path } from 'react-native-svg'
import { useAppTheme } from 'src/app/hooks'
import { Box } from 'src/components/layout/Box'
import { buildGraph } from 'src/components/PriceChart/Model'
import { useHourlyTokenPrices } from 'src/features/historicalChainData/hooks'
import { logger } from 'src/utils/logger'

// number of datapoints to normalize to
const GRAPH_PRECISION = 10
const WIDTH = 45
const HEIGHT = 20

interface InlineGraphProps {
  currency: Currency
}

export const InlinePriceChart = ({ currency }: InlineGraphProps) => {
  const theme = useAppTheme()

  // startOf hour to stabilize the reference
  const yesterday = dayjs().subtract(1, 'day').startOf('hour').unix()

  const { prices, error, loading } = useHourlyTokenPrices({
    token: currency.wrapped,
    periodStartUnix: yesterday,
  })

  if (error) {
    logger.error('InlineGraph', '', `Error fetching price for ${currency.wrapped.symbol}: ${error}`)
  }

  const graph = useMemo(
    () =>
      !error && !loading && prices ? buildGraph(prices, GRAPH_PRECISION, WIDTH, HEIGHT) : undefined,
    [error, loading, prices]
  )

  if (!graph) {
    logger.debug('InlinePriceChart', '', `No price data to display for ${currency.wrapped.symbol}`)
    return null
  }

  return (
    <Box alignSelf="center">
      <Svg height={HEIGHT} width={WIDTH}>
        <Path
          d={serialize(graph.path)}
          stroke={theme.colors.primary1}
          strokeLinecap="square"
          strokeWidth={2}
        />
      </Svg>
    </Box>
  )
}
