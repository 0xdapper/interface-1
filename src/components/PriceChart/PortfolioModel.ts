import { graphql } from 'babel-plugin-relay/macro'
import { logger } from 'ethers'
import { useMemo } from 'react'
import { useLazyLoadQuery } from 'react-relay-offline'
import { GraphMetadatas } from 'src/components/PriceChart/types'
import { buildGraph, GRAPH_PRECISION } from 'src/components/PriceChart/utils'
import { PortfolioModel_PortfolioQuery } from 'src/components/PriceChart/__generated__/PortfolioModel_PortfolioQuery.graphql'

/*
TODO: Added the tokenBalances fields here because usePortfolioBalances's `balancesQuery'
is querying the same endpoint, and somehow, when this portfolioCharts query is run,
react-relay overrides the return value of `balancesQuery` to be undefined. Figure out
how to cache in react-relay properly
*/
const portfolioCharts = graphql`
  query PortfolioModel_PortfolioQuery($ownerAddress: String!) {
    portfolios(ownerAddresses: [$ownerAddress]) {
      hourlyValues: tokensTotalDenominatedValueHistory(duration: HOUR) {
        timestamp
        close: value
      }
      dailyValues: tokensTotalDenominatedValueHistory(duration: DAY) {
        timestamp
        close: value
      }
      weeklyValues: tokensTotalDenominatedValueHistory(duration: WEEK) {
        timestamp
        close: value
      }
      monthlyValues: tokensTotalDenominatedValueHistory(duration: MONTH) {
        timestamp
        close: value
      }
      yearlyValues: tokensTotalDenominatedValueHistory(duration: YEAR) {
        timestamp
        close: value
      }
      tokenBalances {
        quantity
        denominatedValue {
          currency
          value
        }
        token {
          chain
          address
          name
          symbol
          decimals
        }
        tokenProjectMarket {
          tokenProject {
            logoUrl
          }
          relativeChange24: pricePercentChange(duration: DAY) {
            value
          }
        }
      }
    }
  }
`

export function usePortfolioBalanceGraphs(owner: Address): NullUndefined<GraphMetadatas> {
  const { data: portfolioData } = useLazyLoadQuery<PortfolioModel_PortfolioQuery>(portfolioCharts, {
    ownerAddress: owner,
  })

  return useMemo(() => {
    if (!portfolioData) {
      return undefined
    }

    const { hourlyValues, dailyValues, weeklyValues, monthlyValues, yearlyValues } =
      portfolioData.portfolios?.[0] ?? {}

    if (!hourlyValues || !dailyValues || !weeklyValues || !monthlyValues || !yearlyValues) {
      logger.debug('PortfolioModel/usePortfolioBalanceGraphs', 'Historical portfolio prices error')
      return null
    }

    const graphs = [
      {
        label: '1H',
        index: 0,
        data: buildGraph(hourlyValues as any, GRAPH_PRECISION),
      },
      {
        label: '1D',
        index: 1,
        data: buildGraph(dailyValues as any, GRAPH_PRECISION),
      },
      {
        label: '1W',
        index: 2,
        data: buildGraph(weeklyValues as any, GRAPH_PRECISION),
      },
      {
        label: '1M',
        index: 3,
        data: buildGraph(monthlyValues as any, GRAPH_PRECISION),
      },
      {
        label: '1Y',
        index: 4,
        data: buildGraph(yearlyValues as any, GRAPH_PRECISION),
      },
    ] as const

    return graphs as GraphMetadatas
  }, [portfolioData])
}
