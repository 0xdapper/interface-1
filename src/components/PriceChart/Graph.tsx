import React from 'react'
import { StyleSheet, TouchableWithoutFeedback, View } from 'react-native'
import Animated, {
  useAnimatedProps,
  useAnimatedStyle,
  useSharedValue,
  withTiming,
} from 'react-native-reanimated'
import { mixPath, useVector } from 'react-native-redash'
import Svg, { Defs, LinearGradient, Path, Stop } from 'react-native-svg'
import { AnimatedBox, Box } from 'src/components/layout/Box'
import { Cursor } from 'src/components/PriceChart/Cursor'
import { Header } from 'src/components/PriceChart/Header'
import { HEIGHT, NUM_GRAPHS, WIDTH } from 'src/components/PriceChart/Model'
import { GraphMetadatas } from 'src/components/PriceChart/types'
import { Text } from 'src/components/Text'

const AnimatedPath = Animated.createAnimatedComponent(Path)

const SELECTION_WIDTH = WIDTH - 50
const BUTTON_WIDTH = SELECTION_WIDTH / NUM_GRAPHS

interface GraphProps {
  graphs: GraphMetadatas
}

/**
 * Displays a set of graphs with header, scrubbable chart and navigation row.
 * Inspired by https://github.com/wcandillon/can-it-be-done-in-react-native/tree/master/season4/src/Rainbow
 */
export const Graph = ({ graphs }: GraphProps) => {
  const translation = useVector()
  const transition = useSharedValue(0)

  const previousGraphIndex = useSharedValue<number>(0)
  const currentGraphIndex = useSharedValue<number>(0)

  // mixes graph paths on index change
  const graphTransitionAnimatedProps = useAnimatedProps(() => {
    const previousPath = graphs[previousGraphIndex.value].data.path
    const currentPath = graphs[currentGraphIndex.value].data.path
    return {
      //TODO(judo): make animation interruptible
      d: mixPath(transition.value, previousPath, currentPath),
    }
  })
  const graphTransitionClosedAnimatedProps = useAnimatedProps(() => {
    //TODO(judo): figure out a way to merge with `graphTransitionAnimatedProps`
    const previousPath = graphs[previousGraphIndex.value].data.path
    const currentPath = graphs[currentGraphIndex.value].data.path
    return {
      d: `${mixPath(
        transition.value,
        previousPath,
        currentPath
      )} L ${WIDTH} ${HEIGHT} L 0 ${HEIGHT}`,
    }
  })

  const sliderStyle = useAnimatedStyle(() => ({
    transform: [{ translateX: withTiming(BUTTON_WIDTH * currentGraphIndex.value) }],
  }))

  return (
    <Box flex={1}>
      <Header translation={translation} index={currentGraphIndex} graphs={graphs} />
      <View>
        <Svg width={WIDTH} height={HEIGHT}>
          <AnimatedPath
            animatedProps={graphTransitionAnimatedProps}
            fill="transparent"
            stroke="#FF007A"
            strokeWidth={2}
          />

          <Defs>
            <LinearGradient x1="50%" y1="0%" x2="50%" y2="100%" id="gradient">
              <Stop stopColor="#FF007A" offset="0%" stopOpacity="0.2" />
              <Stop stopColor="#FF007A" offset="100%" stopOpacity="0" />
            </LinearGradient>
          </Defs>

          <AnimatedPath
            animatedProps={graphTransitionClosedAnimatedProps}
            fill="url(#gradient)"
            stroke="transparent"
            strokeWidth={3}
          />
        </Svg>
        <Cursor translation={translation} index={currentGraphIndex} graphs={graphs} />
      </View>
      <Box flexDirection="row" width={SELECTION_WIDTH} alignSelf="center">
        <View style={StyleSheet.absoluteFill}>
          <AnimatedBox
            bg="pink"
            borderRadius="lg"
            width={BUTTON_WIDTH}
            style={[StyleSheet.absoluteFillObject, sliderStyle]}
          />
        </View>
        {graphs.map((graph, index) => {
          return (
            <TouchableWithoutFeedback
              key={graph.label}
              onPress={() => {
                previousGraphIndex.value = currentGraphIndex.value
                transition.value = 0
                currentGraphIndex.value = index
                transition.value = withTiming(1)
              }}>
              <AnimatedBox padding="xs" width={BUTTON_WIDTH}>
                <Text
                  variant="buttonLabel"
                  color={index === currentGraphIndex.value ? 'white' : 'primary1'}
                  textAlign="center">
                  {graph.label}
                </Text>
              </AnimatedBox>
            </TouchableWithoutFeedback>
          )
        })}
      </Box>
    </Box>
  )
}
