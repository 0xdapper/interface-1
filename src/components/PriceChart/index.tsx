import { Token } from '@uniswap/sdk-core'
import React from 'react'
import { Graph } from 'src/components/PriceChart/Graph'
import { useGraphs } from 'src/components/PriceChart/Model'
import { Text } from 'src/components/Text'

interface PriceChartProps {
  token: Token
}

export const PriceChart = ({ token }: PriceChartProps) => {
  const graphs = useGraphs(token)

  return graphs && graphs.length ? <Graph graphs={graphs} /> : <Text variant="h2">Loading</Text>
}
