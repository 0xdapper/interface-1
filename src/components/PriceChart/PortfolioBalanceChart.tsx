import React from 'react'
import { useTranslation } from 'react-i18next'
import { NetworkLogo } from 'src/components/CurrencyLogo/NetworkLogo'
import { Flex } from 'src/components/layout/Flex'
import { PriceChart } from 'src/components/PriceChart'
import { usePortfolioBalanceGraphs } from 'src/components/PriceChart/PortfolioModel'
import { Text } from 'src/components/Text'
import { ChainId } from 'src/constants/chains'

export const PortfolioBalanceChart = ({ owner }: { owner: Address }) => {
  const graphs = usePortfolioBalanceGraphs(owner)
  const { t } = useTranslation()

  const error = graphs === null

  if (error) {
    // TODO(MOB-1553): improve portfolio chart error state
    return (
      <Flex centered mx="lg" my="md">
        <Text color="accentFailure" textAlign="center" variant="bodyLarge">
          {t('Could not retrieve historical portfolio balances')}
        </Text>
      </Flex>
    )
  }

  return (
    <PriceChart
      customChartLabel={
        <Flex
          bg="backgroundContainer"
          borderRadius="sm"
          flexDirection="row"
          gap="xxs"
          px="xs"
          py="xxs">
          <NetworkLogo chainId={ChainId.Mainnet} size={16} />
          <Text color="textTertiary" variant="buttonLabelSmall">
            {t('Mainnet balance')}
          </Text>
        </Flex>
      }
      graphs={graphs}
    />
  )
}
