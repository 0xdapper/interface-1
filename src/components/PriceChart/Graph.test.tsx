import { ThemeProvider } from '@shopify/restyle'
import { fireEvent, render } from '@testing-library/react-native'
import React from 'react'
import 'react-native'
import 'react-native-gesture-handler'
import { Graph } from 'src/components/PriceChart/Graph'
import { buildGraph } from 'src/components/PriceChart/Model'
import { theme } from 'src/styles/theme'

const graphDatapoints = buildGraph([
  {
    timestamp: 0,
    close: 100,
    open: 200,
  },
  {
    timestamp: 1,
    close: 200,
    open: 300,
  },
])

const buildGraphMetadata = (index: number) =>
  ({
    label: `Graph ${index}`,
    index,
    data: graphDatapoints,
  } as const)

const graphs = [
  buildGraphMetadata(0),
  buildGraphMetadata(1),
  buildGraphMetadata(2),
  buildGraphMetadata(3),
  buildGraphMetadata(4),
] as const

it('renders correctly', () => {
  const tree = render(
    <ThemeProvider theme={theme}>
      <Graph graphs={graphs} />
    </ThemeProvider>
  ).toJSON()
  expect(tree).toMatchSnapshot()
})

it('navigates between ranges', () => {
  const { getByText } = render(
    <ThemeProvider theme={theme}>
      <Graph graphs={graphs} />
    </ThemeProvider>
  )

  fireEvent.press(getByText(graphs[1].label))
})
