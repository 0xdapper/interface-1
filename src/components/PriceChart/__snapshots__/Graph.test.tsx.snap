// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<View
  style={
    Array [
      Object {
        "display": "flex",
        "flex": 1,
      },
    ]
  }
>
  <View
    style={
      Array [
        Object {
          "display": "flex",
          "flex": 1,
          "justifyContent": "center",
          "padding": 16,
        },
      ]
    }
  >
    <View
      style={
        Array [
          Object {
            "alignItems": "center",
            "display": "flex",
            "flex": 1,
          },
        ]
      }
    >
      <TextInput
        animatedStyle={
          Object {
            "value": Object {},
          }
        }
        collapsable={false}
        editable={false}
        style={
          Object {
            "color": "#27252E",
            "fontFamily": "System",
            "fontSize": 34,
            "fontWeight": "bold",
            "lineHeight": 42.5,
          }
        }
        text="$200"
        underlineColorAndroid="transparent"
        value="$200"
      />
      <TextInput
        animatedStyle={
          Object {
            "value": Object {
              "color": "#DA2D2B",
            },
          }
        }
        collapsable={false}
        editable={false}
        style={
          Object {
            "color": "#DA2D2B",
            "fontFamily": "System",
            "fontSize": 20,
            "fontWeight": "600",
            "lineHeight": 26,
          }
        }
        text="0%"
        underlineColorAndroid="transparent"
        value="0%"
      />
    </View>
  </View>
  <View>
    <RNSVGSvgView
      bbHeight={180}
      bbWidth={750}
      focusable={false}
      height={180}
      style={
        Array [
          Object {
            "backgroundColor": "transparent",
            "borderWidth": 0,
          },
          Object {
            "flex": 0,
            "height": 180,
            "width": 750,
          },
        ]
      }
      width={750}
    >
      <RNSVGGroup>
        <RNSVGPath
          d="M0,0 C0,0 750,180 750,180"
          fill={0}
          propList={
            Array [
              "fill",
              "stroke",
              "strokeWidth",
            ]
          }
          stroke={4294901882}
          strokeWidth={3}
        />
        <RNSVGDefs>
          <RNSVGLinearGradient
            gradient={
              Array [
                0,
                1727987834,
                1,
                15164138,
              ]
            }
            gradientTransform={null}
            gradientUnits={0}
            name="gradient"
            x1="50%"
            x2="50%"
            y1="0%"
            y2="100%"
          />
        </RNSVGDefs>
        <RNSVGPath
          d="M0,0 C0,0 750,180 750,180 L 750 180 L 0 180"
          fill={
            Array [
              1,
              "gradient",
            ]
          }
          propList={
            Array [
              "fill",
              "stroke",
              "strokeWidth",
            ]
          }
          stroke={0}
          strokeWidth={3}
        />
      </RNSVGGroup>
    </RNSVGSvgView>
    <View
      style={
        Array [
          Object {
            "display": "flex",
            "flex": 1,
          },
          Object {
            "bottom": 0,
            "left": 0,
            "position": "absolute",
            "right": 0,
            "top": 0,
          },
        ]
      }
    >
      <View
        animatedStyle={
          Object {
            "value": Object {},
          }
        }
        collapsable={false}
        forwardedRef={[Function]}
        onGestureHandlerEvent={[Function]}
        onGestureHandlerStateChange={[Function]}
        style={
          Object {
            "bottom": 0,
            "display": "flex",
            "left": 0,
            "position": "absolute",
            "right": 0,
            "top": 0,
          }
        }
      >
        <View
          animatedStyle={
            Object {
              "value": Object {
                "transform": Array [
                  Object {
                    "translateX": -25,
                  },
                  Object {
                    "translateY": -25,
                  },
                  Object {
                    "scale": 0,
                  },
                ],
              },
            }
          }
          collapsable={false}
          style={
            Object {
              "alignItems": "center",
              "backgroundColor": "rgba(243, 71, 191, 0.171)",
              "borderRadius": 999999,
              "display": "flex",
              "height": 50,
              "justifyContent": "center",
              "transform": Array [
                Object {
                  "translateX": -25,
                },
                Object {
                  "translateY": -25,
                },
                Object {
                  "scale": 0,
                },
              ],
              "width": 50,
            }
          }
        >
          <View
            style={
              Array [
                Object {
                  "backgroundColor": "#E8006F",
                  "borderRadius": 999999,
                  "display": "flex",
                  "height": 15,
                  "width": 15,
                },
              ]
            }
          />
        </View>
      </View>
    </View>
  </View>
  <View
    style={
      Array [
        Object {
          "alignSelf": "center",
          "display": "flex",
          "flexDirection": "row",
          "width": 718,
        },
      ]
    }
  >
    <View
      style={
        Object {
          "bottom": 0,
          "left": 0,
          "position": "absolute",
          "right": 0,
          "top": 0,
        }
      }
    >
      <View
        animatedStyle={
          Object {
            "value": Object {
              "transform": Array [
                Object {
                  "translateX": 0,
                },
              ],
            },
          }
        }
        collapsable={false}
        style={
          Object {
            "backgroundColor": "#e3e3e3",
            "borderRadius": 6,
            "bottom": 0,
            "display": "flex",
            "left": 0,
            "position": "absolute",
            "right": 0,
            "top": 0,
            "transform": Array [
              Object {
                "translateX": 0,
              },
            ],
            "width": 143.6,
          }
        }
      />
    </View>
    <View
      accessible={true}
      animatedStyle={
        Object {
          "value": Object {},
        }
      }
      collapsable={false}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "display": "flex",
          "padding": 8,
          "width": 143.6,
        }
      }
    >
      <Text
        style={
          Array [
            Object {
              "color": "#E8006F",
              "display": "flex",
              "fontFamily": "System",
              "fontSize": 16,
              "fontWeight": "500",
              "lineHeight": 24,
              "textAlign": "center",
            },
          ]
        }
      >
        Graph 0
      </Text>
    </View>
    <View
      accessible={true}
      animatedStyle={
        Object {
          "value": Object {},
        }
      }
      collapsable={false}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "display": "flex",
          "padding": 8,
          "width": 143.6,
        }
      }
    >
      <Text
        style={
          Array [
            Object {
              "color": "#E8006F",
              "display": "flex",
              "fontFamily": "System",
              "fontSize": 16,
              "fontWeight": "500",
              "lineHeight": 24,
              "textAlign": "center",
            },
          ]
        }
      >
        Graph 1
      </Text>
    </View>
    <View
      accessible={true}
      animatedStyle={
        Object {
          "value": Object {},
        }
      }
      collapsable={false}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "display": "flex",
          "padding": 8,
          "width": 143.6,
        }
      }
    >
      <Text
        style={
          Array [
            Object {
              "color": "#E8006F",
              "display": "flex",
              "fontFamily": "System",
              "fontSize": 16,
              "fontWeight": "500",
              "lineHeight": 24,
              "textAlign": "center",
            },
          ]
        }
      >
        Graph 2
      </Text>
    </View>
    <View
      accessible={true}
      animatedStyle={
        Object {
          "value": Object {},
        }
      }
      collapsable={false}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "display": "flex",
          "padding": 8,
          "width": 143.6,
        }
      }
    >
      <Text
        style={
          Array [
            Object {
              "color": "#E8006F",
              "display": "flex",
              "fontFamily": "System",
              "fontSize": 16,
              "fontWeight": "500",
              "lineHeight": 24,
              "textAlign": "center",
            },
          ]
        }
      >
        Graph 3
      </Text>
    </View>
    <View
      accessible={true}
      animatedStyle={
        Object {
          "value": Object {},
        }
      }
      collapsable={false}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "display": "flex",
          "padding": 8,
          "width": 143.6,
        }
      }
    >
      <Text
        style={
          Array [
            Object {
              "color": "#E8006F",
              "display": "flex",
              "fontFamily": "System",
              "fontSize": 16,
              "fontWeight": "500",
              "lineHeight": 24,
              "textAlign": "center",
            },
          ]
        }
      >
        Graph 4
      </Text>
    </View>
  </View>
</View>
`;
